### PR TITLE
Change CacheControl to Interceptor based Fetcher

### PR DESCRIPTION
**Pull request by BenSchwab**
## Motivation

The interceptor architecture we use for executing a request is loosely similar to the future direction Apollo JS is going: https://github.com/apollographql/apollo-fetch. If I understand correctly, the vision is that Apollo Fetch will make it easy to write fetch policies like `CACHE_FIRST` or more complicated policies like `CACHE_FIRST_IF_NOT_STALE_BUT_IF_NO_NETWORK_ALLOW_STALE`.

However, shipping with every possible fetch strategy is impossible -- thus our current `CacheControl` approach based on a set of enums, could be made more flexible. This PR proposes making `CacheControl` an `ApolloInterceptor` that contains logic of how to fetch the request. (And changes the name to `Fetcher` to better align with JS). 

## Fetcher
- Just an `ApolloInterceptor`
- Uses `FetchOptions` to configure fetching preferences
- Can change `CacheHeaders` and a boolean `fetchFromCache` to control how the data will be fetched
- Thus CACHE_FIRST is implemented by first proceeding the chain with a FetchOptions with `fetchFromCache = true` and catching an exception. If exception it will try to proceed with `fetchFromCache = false`

### Benefits
- `ApolloCacheInterceptor` has less logic - all fetching logic is contained in the `Fetcher`
- Fairly easy to implement a custom Fetcher for specific needs
- In basic use case, API is no more complicated

```
 githuntFeedCall = application.apolloClient()
        .query(feedQuery)
        .fetcher(ApolloFetchers.CACHE_FIRST);
```

## Mutli-response
Notably this PR enables a fetcher to respond multiple times. The rationale behind this is that when you use Apollo with a persistent sql cache, currently the only fetch policy that works is `NETWORK_FIRST` which negates the value of the sql / in memory cache. This is because if you use `CACHE_FIRST`, after the first cache population, the data will never be automatically refreshed. This could be handled with some sort of "staleness" system, but again, this would result in the cache data always being held until it is stale. Rather, it would be useful to allow a request to fulfill from cache first, and then also, fetch updated data from the network.

This does complicate the callback API, particularly with ApolloRx, as we need to know when a request will return no more data (`onCompleted`).

This PR ships with one double response fetch policy `CACHE_AND_NETWORK`.

## Additional work
- [x] Integrate with ApolloSupport
- [x] Add tests
- [x] Add documentation
- [x] Make sure cancel and dispose works properly

Previous discussions: https://airbnb.quip.com/BVbnAztIGoLm

closes https://github.com/apollographql/apollo-android/issues/432

### CODE REVIEW

**Comment by BenSchwab** [comment]
cc @kompfner @martijnwalraven for thoughts from iOS perspective

**Comment by BenSchwab** [comment]
@sav007 @digitalbuddha Let me know what you think about going in this direction, if things seem good from a conceptual perspective, I'll move forward with writing all the tests, documentation, and code cleanup

**Comment by sav007** [comment]
I love this idea, clean and flexible

File: apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/Fetcher.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.apollographql.apollo.fetcher;
+
+import com.apollographql.apollo.interceptor.ApolloInterceptor;
+import com.apollographql.apollo.internal.util.ApolloLogger;
+
+public interface Fetcher {
+
+  ApolloInterceptor create(ApolloLogger logger);
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/Fetcher.java]: should it be renamed for clarity to smth `provideInterceptor`?

File: apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/Fetcher.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.apollographql.apollo.fetcher;
+
+import com.apollographql.apollo.interceptor.ApolloInterceptor;
+import com.apollographql.apollo.internal.util.ApolloLogger;
+
+public interface Fetcher {
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/Fetcher.java]: Should be called `ResponseFetcher`?

**Comment by digitalbuddha** [comment]
I like the concept but am not crazy with the builder method name 

   .fetcher(ApolloFetchers.CACHE_FIRST) feels weird. It's not a fetcher but a cache policy or a fetch policy. When  I see a builder method called fetcher I expect to pass in something that has an implementation for fetch.

**Comment by digitalbuddha** [comment]
Ok next time I'll look at code and not just description of pr. It does fetch! I'm with Ivan love it.

File: apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/CacheAndNetworkFetcher.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  private synchronized void handleNetworkResponse(ApolloInterceptor.InterceptorResponse response) {
+    networkResponse = Optional.of(response);
+    dispatch();
+  }
+
+  private synchronized void handleNetworkError(ApolloException exception) {
+    networkException = Optional.of(exception);
+    dispatch();
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/CacheAndNetworkFetcher.java]: should we consider to not dispatch if result if it was canceled (disposed)

**Comment by sav007** [comment]
We need to make sure that `cancel` on `RealApolloCall` works as expected, all chained interceptors are been disposed

**Comment by BenSchwab** [comment]
Divided this into runtime changes, rx changes, and test changes.

Also, took a stab at https://github.com/apollographql/apollo-android/issues/560 as this PR required me to address cancellation in the new fetchers.

There is a slight change in behavior of cancellation:

1) If cancel is called before `enqueue` or `execute`, there will be an exception
2) If cancel is called during `execute`, there will be an exception
3) If cancel is called during `enqueue` there will be _no_ exception, and callback will not be triggered

The change is in point (3) which would previously trigger onError. This felt weird to me, as if I call `cancel` I don't expect any callbacks (Rx follows this model).

To avoid the memory leaks https://github.com/apollographql/apollo-android/issues/560 during cancelation a made the callback a WeakReference.

I also more aggressively check if dispose has been called on interceptors to avoid doing unnecessary work.

File: apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/ApolloResponseFetcher.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package com.apollographql.apollo.fetcher;
+
+import com.apollographql.apollo.internal.fetcher.CacheFirstFetcher;
+import com.apollographql.apollo.internal.fetcher.CacheOnlyFetcher;
+import com.apollographql.apollo.internal.fetcher.CacheAndNetworkFetcher;
+import com.apollographql.apollo.internal.fetcher.NetworkFirstFetcher;
+import com.apollographql.apollo.internal.fetcher.NetworkOnlyFetcher;
+
+public final class ApolloResponseFetcher {
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/ApolloResponseFetcher.java]: nit: should be named `ApolloResponseFetchers`, like in java `Executors`, in Rx `Schedulers`

File: apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/FetchOptions.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * Immutable options to configure how a {@link ApolloInterceptorChain} will fetch a response.
+ */
+public final class FetchOptions {
+
+  public final boolean fetchFromCache;
+  public final CacheHeaders cacheHeaders;
+
+  public static final FetchOptions NETWORK = new FetchOptions(false, CacheHeaders.NONE);
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/interceptor/FetchOptions.java]: nit: `NETWORK_ONLY`?

**Comment by sav007** [comment]
Just curios if `FetchOptions` is really necessary, is behaviour inside the chained interceptors provided by `ResponseFetcher` is not enough?

**Comment by BenSchwab** [comment]
While, I believe it would be possible to write a custom Fetcher (ApolloInterceptor) that could resolve from network / cache, ideally user's can just proceed with the combo of CacheInterceptor/ParseInterceptor/ServerIntercept that we already assemble. It seemed to me the easiest way to pass a configuration to `CacheInterceptor` to instruct it how to proceed is an options object with `fetchFromCache` and `CacheHeaders` both of which many custom fetch strategies would change to achieve different effects (adding a CacheHeader like "ALLOW_STALE" after an initial cache and network miss)

**Comment by sav007** [comment]
We should update java doc for Callback: `public abstract void onResponse(@Nonnull Response<T> response);` to mention that this can be called multiple times, user should be prepared for this

**Comment by sav007** [comment]
And we need to mention that `Callback.onCompleted` won't be called in case of error

File: apollo-runtime/src/main/java/com/apollographql/apollo/ApolloCall.java [hunk 6]
```diff
@@ -43,15 +43,16 @@
   void enqueue(@Nullable Callback<T> callback);
 
   /**
-   * Sets the {@link CacheControl} strategy for an ApolloCall object.
+   * Sets the {@link ResponseFetcher} strategy for an ApolloCall object.
    *
-   * @param cacheControl the CacheControl strategy to set
+   * @param fetcher the {@link ResponseFetcher} to use.
    * @return The ApolloCall object with the provided CacheControl strategy
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/ApolloCall.java]: `@return The ApolloCall object with the provided CacheControl strategy` -> 
`@return The ApolloCall object with the provided fetch strategy`

File: apollo-runtime/src/main/java/com/apollographql/apollo/ApolloClient.java [hunk 7]
```diff
@@ -334,24 +334,24 @@ public Builder defaultHttpCachePolicy(@Nonnull HttpCachePolicy.Policy cachePolic
...
-    public Builder defaultCacheControl(@Nonnull CacheControl cacheControl) {
-      this.defaultCacheControl = checkNotNull(cacheControl, "cacheControl == null");
+    public Builder defaultCacheHeaders(@Nonnull CacheHeaders cacheHeaders) {
+      this.defaultCacheHeaders = checkNotNull(cacheHeaders, "cacheHeaders == null");
       return this;
     }
 
     /**
-     * Set the default {@link CacheHeaders} strategy that will be passed
+     * Set the default {@link ResponseFetcher} that will create a {@link ResponseFetcher} to be passed
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/apollo/ApolloClient.java]: smth wrong here `ResponseFetcher` will create a `ResponseFetcher`

File: apollo-rx2support/src/main/java/com/apollographql/apollo/rx2/Rx2Apollo.java [hunk 8]
```diff
@@ -66,30 +63,39 @@ private Rx2Apollo() {
   }
 
   /**
-   * Converts an {@link ApolloCall} to a synchronous Single.
+   * Converts an {@link ApolloCall} to an {@link Observable}.
```
Comment (sav007) [inline apollo-rx2support/src/main/java/com/apollographql/apollo/rx2/Rx2Apollo.java]: I guess we should provide more info why it's `Observable` on cache and network fetch strategy example

File: apollo-rxsupport/src/main/java/com/apollographql/apollo/rx/RxApollo.java [hunk 9]
```diff
@@ -81,30 +79,50 @@ private RxApollo() {
   }
 
   /**
-   * Converts an {@link ApolloCall} to a Single.
+   * Converts an {@link ApolloCall} to a Observable.
```
Comment (sav007) [inline apollo-rxsupport/src/main/java/com/apollographql/apollo/rx/RxApollo.java]: the same as above

**Comment by BenSchwab** [comment]
@sav007 I guess I should bump the version number again since this is a breaking change? Not sure we have enough new stuff since that last release to publish an official release

**Comment by sav007** [comment]
but we have't released 0.4.0 yet, why do we need to bump it to 0.5.0?

**Comment by BenSchwab** [comment]
Ok, I wasn't sure if we were trying to keep snapshot builds being semver capatabilble. But if not, let's wait to bump!

**Pull request merged**